File: android_app/app/src/main/java/com/oceansentinels/app/presentation/ui/components/Badges.kt
package com.oceansentinels.app.presentation.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.oceansentinels.app.domain.model.HazardType
import com.oceansentinels.app.domain.model.IncidentStatus
import com.oceansentinels.app.domain.model.UrgencyLevel
import com.oceansentinels.app.domain.model.UserRole
import com.oceansentinels.app.presentation.ui.theme.OceanColors

/**
 * Badge for incident status
 */
@Composable
fun StatusBadge(
    status: IncidentStatus,
    modifier: Modifier = Modifier
) {
    val (color, text, icon) = when (status) {
        IncidentStatus.PENDING -> Triple(OceanColors.Warning, "Pending", Icons.Default.HourglassEmpty)
        IncidentStatus.VERIFIED -> Triple(OceanColors.Info, "Verified", Icons.Default.Verified)
        IncidentStatus.IN_PROGRESS -> Triple(OceanColors.Primary, "In Progress", Icons.Default.LocalShipping)
        IncidentStatus.RESOLVED -> Triple(OceanColors.Success, "Resolved", Icons.Default.CheckCircle)
        IncidentStatus.CLOSED -> Triple(OceanColors.Gray, "Closed", Icons.Default.Close)
    }
    
    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(20.dp),
        color = color.copy(alpha = 0.1f)
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = color,
                modifier = Modifier.size(16.dp)
            )
            Spacer(modifier = Modifier.width(6.dp))
            Text(
                text = text,
                style = MaterialTheme.typography.labelMedium,
                color = color,
                fontWeight = FontWeight.SemiBold
            )
        }
    }
}

/**
 * Badge for urgency level
 */
@Composable
fun UrgencyBadge(
    urgency: UrgencyLevel,
    modifier: Modifier = Modifier
) {
    val (color, text, icon) = when (urgency) {
        UrgencyLevel.LOW -> Triple(OceanColors.Success, "Low", Icons.Default.ArrowDownward)
        UrgencyLevel.MEDIUM -> Triple(OceanColors.Warning, "Medium", Icons.Default.Remove)
        UrgencyLevel.HIGH -> Triple(OceanColors.Orange, "High", Icons.Default.ArrowUpward)
        UrgencyLevel.CRITICAL -> Triple(OceanColors.Danger, "Critical", Icons.Default.PriorityHigh)
    }
    
    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(20.dp),
        color = color.copy(alpha = 0.1f)
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = color,
                modifier = Modifier.size(16.dp)
            )
            Spacer(modifier = Modifier.width(6.dp))
            Text(
                text = text,
                style = MaterialTheme.typography.labelMedium,
                color = color,
                fontWeight = FontWeight.SemiBold
            )
        }
    }
}

/**
 * Badge for hazard type
 */
@Composable
fun HazardTypeBadge(
    hazardType: HazardType,
    modifier: Modifier = Modifier,
    compact: Boolean = false
) {
    val (color, text, icon) = getHazardTypeInfo(hazardType)
    
    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(if (compact) 8.dp else 12.dp),
        color = color.copy(alpha = 0.15f)
    ) {
        Row(
            modifier = Modifier.padding(
                horizontal = if (compact) 8.dp else 12.dp,
                vertical = if (compact) 4.dp else 8.dp
            ),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = color,
                modifier = Modifier.size(if (compact) 14.dp else 18.dp)
            )
            if (!compact) {
                Spacer(modifier = Modifier.width(6.dp))
                Text(
                    text = text,
                    style = MaterialTheme.typography.labelMedium,
                    color = color,
                    fontWeight = FontWeight.Medium
                )
            }
        }
    }
}

/**
 * Get hazard type display info
 */
fun getHazardTypeInfo(hazardType: HazardType): Triple<Color, String, ImageVector> {
    return when (hazardType) {
        HazardType.HIGH_WAVES -> Triple(Color(0xFF1976D2), "High Waves", Icons.Default.Water)
        HazardType.STRONG_CURRENTS -> Triple(Color(0xFF0288D1), "Strong Currents", Icons.Default.Waves)
        HazardType.FLOODING -> Triple(Color(0xFF00897B), "Coastal Flooding", Icons.Default.WaterDrop)
        HazardType.TSUNAMI -> Triple(Color(0xFFD32F2F), "Tsunami Warning", Icons.Default.Warning)
        HazardType.DEBRIS -> Triple(Color(0xFF607D8B), "Debris/Pollution", Icons.Default.Delete)
        HazardType.EROSION -> Triple(Color(0xFF795548), "Coastal Erosion", Icons.Default.Terrain)
        HazardType.STORM -> Triple(Color(0xFF5E35B1), "Storm Alert", Icons.Default.Thunderstorm)
        HazardType.OTHER -> Triple(Color(0xFF9E9E9E), "Other Hazard", Icons.Default.MoreHoriz)
    }
}

/**
 * Badge for user role
 */
@Composable
fun RoleBadge(
    role: UserRole,
    modifier: Modifier = Modifier
) {
    val (color, text, icon) = when (role) {
        UserRole.ADMIN -> Triple(OceanColors.Danger, "Admin", Icons.Default.AdminPanelSettings)
        UserRole.RESCUE_TEAM -> Triple(OceanColors.Primary, "Rescue Team", Icons.Default.LocalHospital)
        UserRole.AUTHORITY -> Triple(OceanColors.Warning, "Authority", Icons.Default.Shield)
        UserRole.PUBLIC -> Triple(OceanColors.Success, "Citizen", Icons.Default.Person)
    }
    
    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(8.dp),
        color = color.copy(alpha = 0.1f)
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 10.dp, vertical = 4.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = color,
                modifier = Modifier.size(14.dp)
            )
            Spacer(modifier = Modifier.width(4.dp))
            Text(
                text = text,
                style = MaterialTheme.typography.labelSmall,
                color = color,
                fontWeight = FontWeight.SemiBold
            )
        }
    }
}

/**
 * Count badge (e.g., for notifications)
 */
@Composable
fun CountBadge(
    count: Int,
    modifier: Modifier = Modifier,
    color: Color = OceanColors.Danger
) {
    if (count > 0) {
        Surface(
            modifier = modifier.size(20.dp),
            shape = RoundedCornerShape(10.dp),
            color = color
        ) {
            Box(contentAlignment = Alignment.Center) {
                Text(
                    text = if (count > 99) "99+" else count.toString(),
                    style = MaterialTheme.typography.labelSmall,
                    color = Color.White,
                    fontWeight = FontWeight.Bold
                )
            }
        }
    }
}

/**
 * Dot indicator (e.g., for online status)
 */
@Composable
fun StatusDot(
    isActive: Boolean,
    modifier: Modifier = Modifier
) {
    Box(
        modifier = modifier
            .size(10.dp)
            .background(
                color = if (isActive) OceanColors.Success else OceanColors.Gray,
                shape = RoundedCornerShape(5.dp)
            )
    )
}

/**
 * Verification badge
 */
@Composable
fun VerifiedBadge(
    modifier: Modifier = Modifier
) {
    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(4.dp),
        color = OceanColors.Success.copy(alpha = 0.1f)
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 6.dp, vertical = 2.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = Icons.Default.Verified,
                contentDescription = "Verified",
                tint = OceanColors.Success,
                modifier = Modifier.size(12.dp)
            )
            Spacer(modifier = Modifier.width(4.dp))
            Text(
                text = "Verified",
                style = MaterialTheme.typography.labelSmall,
                color = OceanColors.Success
            )
        }
    }
}

/**
 * Priority indicator (animated pulse for critical)
 */
@Composable
fun PriorityIndicator(
    urgency: UrgencyLevel,
    modifier: Modifier = Modifier
) {
    val color = when (urgency) {
        UrgencyLevel.LOW -> OceanColors.Success
        UrgencyLevel.MEDIUM -> OceanColors.Warning
        UrgencyLevel.HIGH -> OceanColors.Orange
        UrgencyLevel.CRITICAL -> OceanColors.Danger
    }
    
    Box(
        modifier = modifier
            .width(4.dp)
            .fillMaxHeight()
            .background(color, RoundedCornerShape(2.dp))
    )
}

File: IOS/OceanSentinels/Presentation/UI/Components/Badges.swift
import SwiftUI

// MARK: - Status Badge

/// Badge for incident status.
struct StatusBadge: View {
    let status: IncidentStatus

    private var config: (color: Color, text: String, icon: String) {
        switch status {
        case .pending: (Color.oceanWarning, "Pending", "hourglass")
        case .verified: (Color.oceanInfo, "Verified", "checkmark.seal.fill")
        case .inProgress: (Color.oceanPrimary, "In Progress", "shippingbox.fill")
        case .resolved: (Color.oceanSuccess, "Resolved", "checkmark.circle.fill")
        case .closed: (Color.gray, "Closed", "xmark.circle.fill")
        }
    }

    var body: some View {
        Label(config.text, systemImage: config.icon)
            .font(.caption.weight(.semibold))
            .foregroundStyle(config.color)
            .padding(.horizontal, 12)
            .padding(.vertical, 6)
            .background(config.color.opacity(0.1), in: Capsule())
    }
}

// MARK: - Urgency Badge

/// Badge for urgency level.
struct UrgencyBadge: View {
    let urgency: UrgencyLevel

    private var config: (color: Color, text: String, icon: String) {
        switch urgency {
        case .low: (Color.oceanSuccess, "Low", "arrow.down")
        case .medium: (Color.oceanWarning, "Medium", "minus")
        case .high: (Color.orange, "High", "arrow.up")
        case .critical: (Color.oceanDanger, "Critical", "exclamationmark.2")
        }
    }

    var body: some View {
        Label(config.text, systemImage: config.icon)
            .font(.caption.weight(.semibold))
            .foregroundStyle(config.color)
            .padding(.horizontal, 12)
            .padding(.vertical, 6)
            .background(config.color.opacity(0.1), in: Capsule())
    }
}

// MARK: - Hazard Type Badge

/// Badge for hazard type.
struct HazardTypeBadge: View {
    let hazardType: HazardType
    var compact: Bool = false

    var body: some View {
        let info = Self.hazardInfo(for: hazardType)
        if compact {
            Image(systemName: info.icon)
                .font(.caption2)
                .foregroundStyle(info.color)
                .padding(6)
                .background(info.color.opacity(0.15), in: RoundedRectangle(cornerRadius: 8))
        } else {
            Label(info.text, systemImage: info.icon)
                .font(.caption.weight(.medium))
                .foregroundStyle(info.color)
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(info.color.opacity(0.15), in: RoundedRectangle(cornerRadius: 12))
        }
    }

    static func hazardInfo(for type: HazardType) -> (color: Color, text: String, icon: String) {
        switch type {
        case .highWaves: (Color(hex: 0x1976D2), "High Waves", "water.waves")
        case .strongCurrents: (Color(hex: 0x0288D1), "Strong Currents", "wind")
        case .flooding: (Color(hex: 0x00897B), "Coastal Flooding", "drop.fill")
        case .tsunami: (Color(hex: 0xD32F2F), "Tsunami Warning", "exclamationmark.triangle.fill")
        case .lostVessel: (Color(hex: 0xF57C00), "Lost Vessel", "figure.open.water.swim")
        case .debris: (Color(hex: 0x607D8B), "Debris/Pollution", "trash.fill")
        case .erosion: (Color(hex: 0x795548), "Coastal Erosion", "mountain.2.fill")
        case .storm: (Color(hex: 0x5E35B1), "Storm Alert", "cloud.bolt.fill")
        case .oilSpill: (Color(hex: 0x33691E), "Oil Spill", "drop.triangle.fill")
        case .other: (Color(hex: 0x9E9E9E), "Other Hazard", "ellipsis.circle.fill")
        }
    }
}

// MARK: - Role Badge

/// Badge for user role.
struct RoleBadge: View {
    let role: UserRole

    private var config: (color: Color, text: String, icon: String) {
        switch role {
        case .admin: (Color.oceanDanger, "Admin", "gearshape.fill")
        case .rescueTeam: (Color.oceanPrimary, "Rescue Team", "cross.case.fill")
        case .authority: (Color.oceanWarning, "Authority", "shield.fill")
        case .public: (Color.oceanSuccess, "Citizen", "person.fill")
        }
    }

    var body: some View {
        Label(config.text, systemImage: config.icon)
            .font(.caption2.weight(.semibold))
            .foregroundStyle(config.color)
            .padding(.horizontal, 10)
            .padding(.vertical, 4)
            .background(config.color.opacity(0.1), in: RoundedRectangle(cornerRadius: 8))
    }
}

// MARK: - Count Badge

struct CountBadge: View {
    let count: Int
    var color: Color = .oceanDanger

    var body: some View {
        if count > 0 {
            Text(count > 99 ? "99+" : "\(count)")
                .font(.caption2.weight(.bold))
                .foregroundStyle(.white)
                .frame(width: 20, height: 20)
                .background(color, in: Circle())
        }
    }
}

// MARK: - Status Dot

struct StatusDot: View {
    let isActive: Bool

    var body: some View {
        Circle()
            .fill(isActive ? Color.oceanSuccess : Color.gray)
            .frame(width: 10, height: 10)
    }
}

// MARK: - Priority Indicator

struct PriorityIndicator: View {
    let urgency: UrgencyLevel

    private var color: Color {
        switch urgency {
        case .low: .oceanSuccess
        case .medium: .oceanWarning
        case .high: .orange
        case .critical: .oceanDanger
        }
    }

    var body: some View {
        RoundedRectangle(cornerRadius: 2)
            .fill(color)
            .frame(width: 4)
    }
}
